src/C/sahap-c.c

```diff
@@ -148,6 +152,188 @@ static void InitializeSystem(void) {
+
+    char* kmer = (char*)malloc((k + 1) * sizeof(char));
+    if (!kmer) return NULL;
+
+    int startIdx = r->numSites - k;
+    for (int i = 0; i < k; ++i) {
+        kmer[i] = r->let[startIdx + i];
+    }
+    kmer[k] = '\0';
+
+    return kmer;
+}
+
+// Extract k-mer from the head of a read (null-terminated)
+char* headKmer(const READ* r, int k) {
+    if (!r || !r->let || k > r->numSites) return NULL;
+
+    char* kmer = (char*)malloc((k + 1) * sizeof(char));
+    if (!kmer) return NULL;
+
+    for (int i = 0; i < k; ++i) {
+        kmer[i] = r->let[i];
+    }
+    kmer[k] = '\0';
+
+    return kmer;
+}
+
+// Check if two k-mers match (string compare)
+bool kmersMatch(const char* kmer1, const char* kmer2) {
+    if (!kmer1 || !kmer2) return false;
+    return strcmp(kmer1, kmer2) == 0;
+}
+
+// Compare two reads for equality
+bool readsEqual(const READ* a, const READ* b) {
+    if (a->numSites != b->numSites) return false;
+    if (a->firstSite != b->firstSite) return false;
+    for (int i = 0; i < a->numSites; ++i) {
+        if (a->let[i] != b->let[i]) return false;
+    }
+    return true;
+}
+
+// Append one base to the right of a read
+READ addReadsRight(const READ* base, char a) {
+    READ out = *base; 
+    out.numSites = base->numSites + 1;
+
+    char* newLet = (char*)malloc(out.numSites * sizeof(char));
+    if (!newLet) {
+        out.let = NULL;
+        return out;
+    }
+
+    for (int i = 0; i < base->numSites; ++i) {
+        newLet[i] = base->let[i];
+    }
+    newLet[out.numSites - 1] = a; 
+
+    out.let = newLet;
+    // firstSite stays the same when extending right
+
+    return out;
+}
+
+// Prepend one base to the left of a read
+READ addReadsLeft(const READ* base, char a) {
+    READ out = *base;
+    out.numSites = base->numSites + 1;
+    out.firstSite = base->firstSite - 1;
+
+    char* newLet = (char*)malloc(out.numSites * sizeof(char));
+    if (!newLet) {
+        out.let = NULL;
+        return out;
+    }
+
+    newLet[0] = a;
+    for (int i = 0; i < base->numSites; ++i) {
+        newLet[i + 1] = base->let[i];
+    }
+
+    out.let = newLet;
+    return out;
+}
+
+// Extend reads by matching k-mers and adding the next base from matching read
+void ExtendReadsUsingKmers(int k) {
+    bool extendedAny = true;
+
+    while (extendedAny) {
+        extendedAny = false;
+
+        for (int r = 0; r < _numReads; ++r) {
+            READ* current = &_read[r];
+            bool extendedThisRead = false;
+
+            // Extend RIGHT: get tail k-mer of current read
+            char* tailK = tailKmer(current, k);
+            if (!tailK) continue;
+
+            for (int j = 0; j < _numReads; ++j) {
+                if (j == r) continue;
+
+                READ* candidate = &_read[j];
+                char* headK = headKmer(candidate, k);
+                if (!headK) continue;
+
+                // If tail of current matches head of candidate, try to extend
+                if (kmersMatch(tailK, headK)) {
```

Comment:
Um... there's no guarantee the end of one matches the opposite end of the other. It's the wrong place to look.

I'll stop here and finish the meeting with a brief call with you guys.